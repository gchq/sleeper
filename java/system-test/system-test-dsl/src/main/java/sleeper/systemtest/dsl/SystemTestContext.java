/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.dsl;

import sleeper.systemtest.dsl.instance.DeployedSleeperInstances;
import sleeper.systemtest.dsl.instance.DeployedSystemTestResources;
import sleeper.systemtest.dsl.instance.SystemTestDeploymentContext;
import sleeper.systemtest.dsl.instance.SystemTestInstanceContext;
import sleeper.systemtest.dsl.instance.SystemTestParameters;
import sleeper.systemtest.dsl.reporting.ReportingContext;
import sleeper.systemtest.dsl.sourcedata.IngestSourceFilesContext;
import sleeper.systemtest.dsl.util.TestContext;

/**
 * Tracks the context of a single running system test that will use the DSL. This is anything that needs to be
 * remembered from one step to the next, but not in between tests.
 * <p>
 * The {@link SleeperSystemTest} DSL uses this for the context of a test. Deployed resources and Sleeper instances are
 * managed separately in {@link SystemTestDeploymentContext}, outside of any test.
 */
public class SystemTestContext {
    private final SystemTestParameters parameters;
    private final DeployedSystemTestResources systemTestResources;
    private final SystemTestInstanceContext instance;
    private final IngestSourceFilesContext sourceFiles;
    private final ReportingContext reporting;

    public SystemTestContext(
            SystemTestParameters parameters, SystemTestDrivers drivers,
            DeployedSystemTestResources systemTestResources, DeployedSleeperInstances deployedInstances,
            TestContext testContext) {
        this.parameters = parameters;
        this.systemTestResources = systemTestResources;
        instance = new SystemTestInstanceContext(parameters, deployedInstances, drivers.instance(parameters), testContext);
        sourceFiles = new IngestSourceFilesContext(systemTestResources, instance);
        reporting = new ReportingContext(parameters);
    }

    public SystemTestParameters parameters() {
        return parameters;
    }

    public DeployedSystemTestResources systemTest() {
        return systemTestResources;
    }

    public SystemTestInstanceContext instance() {
        return instance;
    }

    public IngestSourceFilesContext sourceFiles() {
        return sourceFiles;
    }

    public ReportingContext reporting() {
        return reporting;
    }
}
