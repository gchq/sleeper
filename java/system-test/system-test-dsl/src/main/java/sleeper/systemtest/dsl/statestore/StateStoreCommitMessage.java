/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.dsl.statestore;

import sleeper.core.statestore.FileReference;

import java.util.List;

public class StateStoreCommitMessage {

    private final String tableId;
    private final String body;

    private StateStoreCommitMessage(String tableId, String body) {
        this.tableId = tableId;
        this.body = body;
    }

    public static StateStoreCommitMessage tableIdAndBody(String tableId, String body) {
        return new StateStoreCommitMessage(tableId, body);
    }

    public static Commit addFileWithJob(FileReference file) {
        return factory -> factory.addFileWithJob(file);
    }

    public static Commit addFiles(List<FileReference> files) {
        return factory -> factory.addFiles(files);
    }

    public static Commit addFile(FileReference file) {
        return factory -> factory.addFiles(List.of(file));
    }

    public static Commit addPartitionFile(String partitionId, String filename, long records) {
        return factory -> factory.addFiles(
                List.of(FileReference.builder()
                        .partitionId(partitionId)
                        .filename(filename)
                        .numberOfRecords(records)
                        .countApproximate(false)
                        .onlyContainsDataForThisPartition(true)
                        .build()));
    }

    public String getTableId() {
        return tableId;
    }

    public String getBody() {
        return body;
    }

    public interface Commit {
        StateStoreCommitMessage createMessage(StateStoreCommitMessageFactory factory);
    }

}
