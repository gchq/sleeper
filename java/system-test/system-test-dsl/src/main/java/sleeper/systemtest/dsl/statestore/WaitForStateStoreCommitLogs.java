/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.dsl.statestore;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.core.util.PollWithRetries;

import java.time.Duration;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

public class WaitForStateStoreCommitLogs {
    public static final Logger LOGGER = LoggerFactory.getLogger(WaitForStateStoreCommitLogs.class);
    private static final Duration QUERY_RUNS_TIME_SLACK = Duration.ofSeconds(5);

    private final StateStoreCommitterLogsDriver driver;

    public WaitForStateStoreCommitLogs(StateStoreCommitterLogsDriver driver) {
        this.driver = driver;
    }

    public StateStoreCommitterLogs waitForCommitLogs(PollWithRetries poll, Map<String, Integer> waitForNumCommitsByTableId, Instant getRunsAfterTime) throws InterruptedException {
        LOGGER.info("Waiting for commits by table ID: {}", waitForNumCommitsByTableId);
        Instant startTime = getRunsAfterTime.minus(QUERY_RUNS_TIME_SLACK);
        return poll.queryUntil("all state store commits are applied", () -> {
            Instant endTime = Instant.now().plus(Duration.ofMinutes(10));
            return driver.getLogsInPeriod(startTime, endTime);
        }, logs -> getRemainingCommits(waitForNumCommitsByTableId, logs).isEmpty());
    }

    public static Map<String, Integer> getRemainingCommits(Map<String, Integer> waitForNumCommitsByTableId, StateStoreCommitterLogs logs) {
        Map<String, Integer> numCommitsByTableId = logs.countNumCommitsByTableId(waitForNumCommitsByTableId.keySet());
        Map<String, Integer> remainingCommitsByTableId = countRemainingCommitsByTableId(waitForNumCommitsByTableId, numCommitsByTableId);
        LOGGER.info("Remaining unapplied commits by table ID: {}", remainingCommitsByTableId);
        return remainingCommitsByTableId;
    }

    private static Map<String, Integer> countRemainingCommitsByTableId(
            Map<String, Integer> waitForNumCommitsByTableId, Map<String, Integer> numCommitsByTableId) {
        Map<String, Integer> remainingCommitsByTableId = new HashMap<>(waitForNumCommitsByTableId);
        numCommitsByTableId.forEach((tableId, numCommits) -> {
            remainingCommitsByTableId.compute(tableId, (id, count) -> {
                if (count == null) {
                    return null;
                } else if (numCommits >= count) {
                    return null;
                } else {
                    return count - numCommits;
                }
            });
        });
        return remainingCommitsByTableId;
    }
}
