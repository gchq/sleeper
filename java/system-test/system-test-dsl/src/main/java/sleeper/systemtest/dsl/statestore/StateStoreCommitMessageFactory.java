/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.dsl.statestore;

import sleeper.core.statestore.FileReference;
import sleeper.ingest.job.IngestJob;
import sleeper.ingest.job.commit.IngestAddFilesCommitRequest;
import sleeper.ingest.job.commit.IngestAddFilesCommitRequestSerDe;

import java.time.Instant;
import java.util.List;
import java.util.UUID;
import java.util.function.Consumer;

public class StateStoreCommitMessageFactory {

    private final String tableId;

    public StateStoreCommitMessageFactory(String tableId) {
        this.tableId = tableId;
    }

    public StateStoreCommitMessage addFiles(List<FileReference> files) {
        return ingest(builder -> builder.fileReferences(files));
    }

    public StateStoreCommitMessage addFileWithJob(FileReference file) {
        String jobId = UUID.randomUUID().toString();
        IngestJob job = IngestJob.builder()
                .id(jobId)
                .tableId(tableId)
                .files(List.of("test-file-" + jobId + ".parquet"))
                .build();
        return ingest(builder -> builder
                .fileReferences(List.of(file))
                .ingestJob(job)
                .taskId(jobId)
                .jobRunId(jobId)
                .writtenTime(Instant.now()));
    }

    private StateStoreCommitMessage ingest(Consumer<IngestAddFilesCommitRequest.Builder> config) {
        IngestAddFilesCommitRequest.Builder builder = IngestAddFilesCommitRequest.builder().tableId(tableId);
        config.accept(builder);
        return StateStoreCommitMessage.tableIdAndBody(tableId,
                new IngestAddFilesCommitRequestSerDe().toJson(builder.build()));
    }

}
