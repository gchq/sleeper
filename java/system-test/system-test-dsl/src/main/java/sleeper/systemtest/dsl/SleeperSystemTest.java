/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.dsl;

import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.properties.table.TableProperty;
import sleeper.core.properties.validation.OptionalStack;
import sleeper.core.record.Record;
import sleeper.core.schema.Schema;
import sleeper.systemtest.dsl.compaction.SystemTestCompaction;
import sleeper.systemtest.dsl.gc.SystemTestGarbageCollection;
import sleeper.systemtest.dsl.ingest.SystemTestIngest;
import sleeper.systemtest.dsl.instance.SystemTestInstanceConfiguration;
import sleeper.systemtest.dsl.instance.SystemTestOptionalStacks;
import sleeper.systemtest.dsl.instance.SystemTestParameters;
import sleeper.systemtest.dsl.instance.SystemTestTableFiles;
import sleeper.systemtest.dsl.instance.SystemTestTables;
import sleeper.systemtest.dsl.metrics.SystemTestTableMetrics;
import sleeper.systemtest.dsl.partitioning.SystemTestPartitioning;
import sleeper.systemtest.dsl.python.SystemTestPythonApi;
import sleeper.systemtest.dsl.query.SystemTestQuery;
import sleeper.systemtest.dsl.reporting.SystemTestReporting;
import sleeper.systemtest.dsl.sourcedata.GenerateNumberedRecords;
import sleeper.systemtest.dsl.sourcedata.GenerateNumberedValueOverrides;
import sleeper.systemtest.dsl.sourcedata.RecordNumbers;
import sleeper.systemtest.dsl.sourcedata.SystemTestCluster;
import sleeper.systemtest.dsl.sourcedata.SystemTestLocalFiles;
import sleeper.systemtest.dsl.sourcedata.SystemTestSourceFiles;
import sleeper.systemtest.dsl.statestore.SystemTestStateStore;

import java.nio.file.Path;
import java.util.Collection;
import java.util.Map;
import java.util.stream.LongStream;

import static sleeper.core.properties.table.TableProperty.TABLE_ONLINE;

/**
 * The entry point that all system tests use to interact with the system. This is the starting point for all steps of
 * the Domain Specific Language (DSL) for system tests. The purpose of this is to make it as easy as possible to read
 * and write system tests. It should make it easy to find and reuse steps to interact with the systems you care about
 * for your tests.
 * <p>
 * It does this by delegating to drivers to actually interact with the system. The DSL only defines
 * the steps and language to be used by tests, and not any of the implemented behaviour of the steps.
 * <p>
 * This class should match as closely as possible with a diagram of the components of the system. You can
 * expect to find the system you care about through a method on this class. Some core features of the system are
 * on this class directly, but we try to limit its size by grouping methods into components which you access through
 * a method, eg. {@link #ingest()}.
 * <p>
 * Most tests will use steps from different systems, but where multiple steps run against the same system in succession
 * we can use method chaining to avoid accessing that system repeatedly. If you go from one system to another and back,
 * assume you should re-access the first system from here again the second time.
 * Try to avoid assigning variables except for data you want to reuse.
 */
public class SleeperSystemTest {

    private final SystemTestParameters parameters;
    private final SystemTestDrivers baseDrivers;
    private final SystemTestContext context;

    public SleeperSystemTest(SystemTestParameters parameters, SystemTestDrivers baseDrivers, SystemTestContext context) {
        this.parameters = parameters;
        this.baseDrivers = baseDrivers;
        this.context = context;
    }

    public void connectToInstanceAddOnlineTable(SystemTestInstanceConfiguration configuration) {
        context.instance().connectTo(configuration);
        context.instance().addDefaultTables(Map.of(TABLE_ONLINE, "true"));
    }

    public void connectToInstanceAddOfflineTable(SystemTestInstanceConfiguration configuration) {
        context.instance().connectTo(configuration);
        context.instance().addDefaultTables(Map.of(TABLE_ONLINE, "false"));
    }

    public void connectToInstanceNoTables(SystemTestInstanceConfiguration configuration) {
        context.instance().connectTo(configuration);
    }

    public InstanceProperties instanceProperties() {
        return context.instance().getInstanceProperties();
    }

    public TableProperties tableProperties() {
        return context.instance().getTableProperties();
    }

    public void updateTableProperties(Map<TableProperty, String> values) {
        context.instance().updateTableProperties(values);
    }

    public SystemTestSourceFiles sourceFiles() {
        return new SystemTestSourceFiles(context.instance(), context.sourceFiles(), baseDrivers.sourceFiles(context));
    }

    public SystemTestTableFiles tableFiles() {
        return new SystemTestTableFiles(context.instance());
    }

    public SystemTestPartitioning partitioning() {
        return new SystemTestPartitioning(context);
    }

    public SystemTestIngest ingest() {
        return new SystemTestIngest(context, baseDrivers);
    }

    public SystemTestQuery query() {
        return new SystemTestQuery(context, baseDrivers);
    }

    public SystemTestQuery directQuery() {
        return query().direct();
    }

    public SystemTestCompaction compaction() {
        return new SystemTestCompaction(context, baseDrivers);
    }

    public SystemTestGarbageCollection garbageCollection() {
        return new SystemTestGarbageCollection(context);
    }

    public SystemTestReporting reporting() {
        return new SystemTestReporting(context);
    }

    public SystemTestTableMetrics tableMetrics() {
        return new SystemTestTableMetrics(context.instance().adminDrivers().tableMetrics(context));
    }

    public SystemTestCluster systemTestCluster() {
        return new SystemTestCluster(context, baseDrivers);
    }

    public SystemTestPythonApi pythonApi() {
        return new SystemTestPythonApi(context);
    }

    public SystemTestLocalFiles localFiles(Path tempDir) {
        return new SystemTestLocalFiles(context.instance(), baseDrivers.localFiles(context), tempDir);
    }

    public void setGeneratorOverrides(GenerateNumberedValueOverrides overrides) {
        context.instance().setGeneratorOverrides(overrides);
    }

    public GenerateNumberedRecords numberedRecords() {
        return context.instance().numberedRecords();
    }

    public Iterable<Record> generateNumberedRecords(LongStream numbers) {
        return context.instance().numberedRecords().iterableFrom(numbers);
    }

    public Iterable<Record> generateNumberedRecords(Schema schema, LongStream numbers) {
        return context.instance().numberedRecords(schema).iterableFrom(numbers);
    }

    public RecordNumbers scrambleNumberedRecords(LongStream longStream) {
        return RecordNumbers.scrambleNumberedRecords(longStream);
    }

    public Path getSplitPointsDirectory() {
        return parameters.getScriptsDirectory()
                .resolve("test/splitpoints");
    }

    public void enableOptionalStack(OptionalStack stack) {
        new SystemTestOptionalStacks(context.instance()).addOptionalStack(stack);
    }

    public void enableOptionalStacks(Collection<OptionalStack> stacks) {
        new SystemTestOptionalStacks(context.instance()).addOptionalStacks(stacks);
    }

    public void disableOptionalStack(OptionalStack stack) {
        new SystemTestOptionalStacks(context.instance()).removeOptionalStack(stack);
    }

    public void disableOptionalStacks(Collection<OptionalStack> stacks) {
        new SystemTestOptionalStacks(context.instance()).removeOptionalStacks(stacks);
    }

    public SystemTestStateStore stateStore() {
        return new SystemTestStateStore(context);
    }

    public SystemTestTables tables() {
        return new SystemTestTables(context.instance());
    }

    public SleeperSystemTest table(String name) {
        context.instance().setCurrentTable(name);
        return this;
    }
}
