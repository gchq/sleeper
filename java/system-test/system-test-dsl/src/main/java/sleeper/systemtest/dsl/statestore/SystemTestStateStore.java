/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.dsl.statestore;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.core.table.TableStatus;
import sleeper.systemtest.dsl.SystemTestContext;
import sleeper.systemtest.dsl.SystemTestDrivers;
import sleeper.systemtest.dsl.instance.SystemTestInstanceContext;

import java.time.Instant;
import java.util.Map;

import static java.util.stream.Collectors.toMap;
import static java.util.stream.Collectors.toSet;
import static sleeper.configuration.properties.table.TableProperty.TABLE_ID;

public class SystemTestStateStore {
    public static final Logger LOGGER = LoggerFactory.getLogger(SystemTestStateStore.class);

    private final SystemTestContext context;
    private final StateStoreCommitterDriver driver;
    private final StateStoreCommitterLogsDriver logsDriver;

    public SystemTestStateStore(SystemTestContext context) {
        this.context = context;
        SystemTestDrivers adminDrivers = context.instance().adminDrivers();
        driver = adminDrivers.stateStoreCommitter(context);
        logsDriver = adminDrivers.stateStoreCommitterLogs(context);
    }

    public SystemTestStateStoreFakeCommits fakeCommits() {
        return new SystemTestStateStoreFakeCommits(context, driver, logsDriver);
    }

    public double commitsPerSecondForTable() {
        TableStatus table = context.instance().getTableStatus();
        double commitsPerSecond = commitsPerSecondByTableId()
                .getOrDefault(table.getTableUniqueId(), 0.0);
        LOGGER.info("Found commits per second for table {}: {}", table, commitsPerSecond);
        return commitsPerSecond;
    }

    public Map<String, Double> commitsPerSecondByTable() {
        Map<String, Double> byTableId = commitsPerSecondByTableId();
        SystemTestInstanceContext instance = context.instance();
        return instance.streamTableProperties()
                .collect(toMap(
                        table -> instance.getTestTableName(table),
                        table -> {
                            Double commitsPerSecond = byTableId.getOrDefault(table.get(TABLE_ID), 0.0);
                            LOGGER.info("Found commits per second for table {}: {}", table.getStatus(), commitsPerSecond);
                            return commitsPerSecond;
                        }));
    }

    private Map<String, Double> commitsPerSecondByTableId() {
        return logsDriver.getLogsInPeriod(context.reporting().getRecordingStartTime(), Instant.now())
                .computeOverallCommitsPerSecondByTableId(
                        context.instance().streamTableProperties()
                                .map(table -> table.get(TABLE_ID))
                                .collect(toSet()));
    }

}
