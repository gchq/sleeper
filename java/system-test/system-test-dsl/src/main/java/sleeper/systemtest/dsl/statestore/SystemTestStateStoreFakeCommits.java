/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.dsl.statestore;

import sleeper.core.util.PollWithRetries;
import sleeper.systemtest.dsl.SystemTestContext;
import sleeper.systemtest.dsl.instance.SystemTestInstanceContext;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toUnmodifiableList;
import static sleeper.configuration.properties.table.TableProperty.TABLE_ID;

public class SystemTestStateStoreFakeCommits {

    private final SystemTestInstanceContext instance;
    private final StateStoreCommitterDriver driver;
    private final WaitForStateStoreCommitLogs waiter;
    private final Map<String, Integer> waitForNumCommitsByTableId = new ConcurrentHashMap<>();
    private final Instant getRunsAfterTime;

    public SystemTestStateStoreFakeCommits(
            SystemTestContext context,
            StateStoreCommitterDriver driver,
            StateStoreCommitterLogsDriver logsDriver) {
        this.driver = driver;
        instance = context.instance();
        waiter = new WaitForStateStoreCommitLogs(logsDriver);
        getRunsAfterTime = context.reporting().getRecordingStartTime();
    }

    public SystemTestStateStoreFakeCommits sendBatched(Stream<StateStoreCommitMessage.Commit> commits) {
        StateStoreCommitMessageFactory factory = messageFactory();
        send(commits.map(commit -> commit.createMessage(factory)));
        return this;
    }

    public SystemTestStateStoreFakeCommits sendBatchedForEachTable(Stream<StateStoreCommitMessage.Commit> commits) {
        List<StateStoreCommitMessageFactory> factories = instance.streamTableProperties()
                .map(table -> table.get(TABLE_ID))
                .map(StateStoreCommitMessageFactory::new)
                .collect(toUnmodifiableList());
        send(commits.flatMap(commit -> factories.stream().map(factory -> commit.createMessage(factory))));
        return this;
    }

    public SystemTestStateStoreFakeCommits send(StateStoreCommitMessage.Commit commit) {
        send(Stream.of(commit.createMessage(messageFactory())));
        return this;
    }

    public SystemTestStateStoreFakeCommits waitForCommitLogs(PollWithRetries poll) throws InterruptedException {
        waiter.waitForCommitLogs(poll, waitForNumCommitsByTableId, getRunsAfterTime);
        return this;
    }

    public SystemTestStateStoreFakeCommits pauseReceivingCommitMessages() {
        driver.pauseReceivingMessages();
        return this;
    }

    public SystemTestStateStoreFakeCommits resumeReceivingCommitMessages() {
        driver.resumeReceivingMessages();
        return this;
    }

    private void send(Stream<StateStoreCommitMessage> messages) {
        driver.sendCommitMessages(messages
                .peek(message -> waitForNumCommitsByTableId.compute(
                        message.getTableId(),
                        (id, count) -> count == null ? 1 : count + 1)));
    }

    private StateStoreCommitMessageFactory messageFactory() {
        return new StateStoreCommitMessageFactory(instance.getTableStatus().getTableUniqueId());
    }
}
