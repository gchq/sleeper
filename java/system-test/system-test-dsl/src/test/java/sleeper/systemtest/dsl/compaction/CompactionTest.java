/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.dsl.compaction;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import sleeper.compaction.core.job.creation.strategy.impl.BasicCompactionStrategy;
import sleeper.core.partition.PartitionsBuilder;
import sleeper.core.record.testutils.SortedRecordsCheck;
import sleeper.core.statestore.AllReferencesToAllFiles;
import sleeper.systemtest.dsl.SleeperSystemTest;
import sleeper.systemtest.dsl.testutil.InMemoryDslTest;

import java.nio.file.Path;
import java.util.Map;
import java.util.stream.LongStream;

import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.core.properties.model.IngestFileWritingStrategy.ONE_FILE_PER_LEAF;
import static sleeper.core.properties.model.IngestFileWritingStrategy.ONE_REFERENCE_PER_LEAF;
import static sleeper.core.properties.table.TableProperty.COMPACTION_FILES_BATCH_SIZE;
import static sleeper.core.properties.table.TableProperty.COMPACTION_STRATEGY_CLASS;
import static sleeper.core.properties.table.TableProperty.INGEST_FILE_WRITING_STRATEGY;
import static sleeper.core.testutils.printers.FileReferencePrinter.printFiles;
import static sleeper.systemtest.dsl.sourcedata.GenerateNumberedValue.addPrefix;
import static sleeper.systemtest.dsl.sourcedata.GenerateNumberedValue.numberStringAndZeroPadTo;
import static sleeper.systemtest.dsl.sourcedata.GenerateNumberedValueOverrides.overrideField;
import static sleeper.systemtest.dsl.testutil.InMemoryTestInstance.IN_MEMORY_MAIN;
import static sleeper.systemtest.dsl.util.SystemTestSchema.DEFAULT_SCHEMA;
import static sleeper.systemtest.dsl.util.SystemTestSchema.ROW_KEY_FIELD_NAME;

@InMemoryDslTest
public class CompactionTest {
    private final Path tempDir = null;

    @BeforeEach
    void setUp(SleeperSystemTest sleeper) throws Exception {
        sleeper.connectToInstanceNoTables(IN_MEMORY_MAIN);
        sleeper.tables().createWithProperties("compaction", DEFAULT_SCHEMA, Map.of(
                COMPACTION_STRATEGY_CLASS, BasicCompactionStrategy.class.getName(),
                COMPACTION_FILES_BATCH_SIZE, "2"));
    }

    @Test
    void shouldCompactFilesFromMultiplePartitions(SleeperSystemTest sleeper) throws Exception {
        // Given we have 4 leaf partitions, LL, LR, RL, RR
        sleeper.setGeneratorOverrides(overrideField(
                ROW_KEY_FIELD_NAME, numberStringAndZeroPadTo(2).then(addPrefix("row-"))));
        sleeper.partitioning().setPartitions(new PartitionsBuilder(DEFAULT_SCHEMA)
                .rootFirst("root")
                .splitToNewChildren("root", "L", "R", "row-50")
                .splitToNewChildren("L", "LL", "LR", "row-25")
                .splitToNewChildren("R", "RL", "RR", "row-75")
                .buildTree());
        // And half the partitions have a file A wholly on each partition
        sleeper.updateTableProperties(Map.of(INGEST_FILE_WRITING_STRATEGY, ONE_FILE_PER_LEAF.toString()));
        sleeper.ingest().direct(tempDir)
                .numberedRecords(LongStream.range(0, 50));
        // And we have a file B containing data for all partitions, referenced on each
        sleeper.updateTableProperties(Map.of(INGEST_FILE_WRITING_STRATEGY, ONE_REFERENCE_PER_LEAF.toString()));
        sleeper.ingest().direct(tempDir)
                .numberedRecords(sleeper.scrambleNumberedRecords(LongStream.range(0, 100)).stream());
        // And we have a file C in the root partition
        sleeper.sourceFiles().inDataBucket().writeSketches()
                .createWithNumberedRecords("file.parquet", LongStream.range(50, 100));
        sleeper.ingest().toStateStore().addFileOnPartition("file.parquet", "root", 50);

        // When
        sleeper.compaction()
                // Merge files A and B on LL and LR, split C one level down to L and R
                .createJobs(2).waitForTasks(1).waitForJobs()
                // Split C another level to LL and LR, merge it with the existing data
                .createJobs(4).waitForTasks(1).waitForJobs();

        // Then
        assertThat(sleeper.directQuery().allRecordsInTable())
                .containsExactlyInAnyOrderElementsOf(sleeper.generateNumberedRecords(
                        LongStream.range(0, 100).flatMap(number -> LongStream.of(number, number))));
        AllReferencesToAllFiles files = sleeper.tableFiles().all();
        Approvals.verify(printFiles(sleeper.partitioning().tree(), files));
        assertThat(files.getFilesWithReferences())
                .allSatisfy(file -> assertThat(
                        SortedRecordsCheck.check(DEFAULT_SCHEMA, sleeper.getRecords(file)))
                        .isEqualTo(SortedRecordsCheck.sorted(file.estimateRecords())));
    }

}
