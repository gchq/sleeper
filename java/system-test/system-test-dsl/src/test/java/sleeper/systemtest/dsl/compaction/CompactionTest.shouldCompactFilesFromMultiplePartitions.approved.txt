Unreferenced files: 6
Referenced files: 4
File references: 4
Partition at LL: 50 records in file 1
Partition at LR: 50 records in file 2
Partition at RL: 50 records in file 3
Partition at RR: 50 records in file 4
